### PR TITLE
Create simple version of ratelimit package

### PR DESCRIPTION
**Pull request by justinsb**
Allows for better testing.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35583)

### CODE REVIEW

**Comment by justinsb** [comment]
@k8s-bot cvm gce e2e test this

**Comment by justinsb** [comment]
@k8s-bot gci gce e2e test this

**Comment by justinsb** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by jbeda** [comment]
Awesome!  Happy to review this.

Quick request -- can you break out the godep/vendor stuff in to a separate commit?  It helps to keep the "real code" easy to review.

**Review by jbeda — COMMENTED** [review]
Looks pretty good.  Mostly nits around comments.

File: pkg/util/ratelimit/bucket.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+import (
+	"sync"
+	"time"
+)
+
+// Bucket models a token bucket
+type Bucket struct {
+	unitsPerNano float64
+	nanosPerUnit float64
```
Comment (justinsb) [inline pkg/util/ratelimit/bucket.go]: Will do - I also hand-waving satisifed myself that there's no loss of precision for qps from about 1E6 to 1E-6.

I'll actually add some unit tests just to prove it :-)

**Comment by justinsb** [comment]
The "just one more unit test" found a bug actually ... pathological rates like (1E9 + 1) were a bit tricky, so I added a "fractional capacity" field which holds the non-integer capacity we added, which makes the time math much simpler.

**Comment by justinsb** [comment]
@k8s-bot node e2e test this

**Comment by luxas** [comment]
Oh, how sweet isn't this?
Removing 1k LoC is always fun :)
:+1:

Will we use this ratelimit package when we're gonna limit the unauthorized access to the `kube-public` namespace?

**Comment by jbeda** [comment]
I'm really not sure why this hasn't been merged.  Let me toggle LGTM to see if that does it.

**Comment by luxas** [comment]
It's just because the Submit-Queue is sooo long unfortunately.

If this is urgent, add priority/P2 and it will merge quite soon

**Comment by smarterclayton** [comment]
Changes like this should have included review from sig-API-machinery

**Comment by mikedanese** [comment]
FYI, there is a ratelimiter in [golang.org/x/time](https://godoc.org/golang.org/x/time/rate).

**Comment by liggitt** [comment]
not opposed to switching if there's a shortcoming of the one in use, but stuff under golang.org/x/... still has to be vendored in, and that once doesn't appear significantly simpler or more capable for our needs

**Comment by mikedanese** [comment]
From my understanding, the main problems we've (google) had around the juju ratelimiter is the license. We had to hack around it to be able to use the client internally because lawyers. I don't understand the "allows for better testing" but would like to understand the juju ratelimter short-comings better.

**Comment by liggitt** [comment]
looks like golang.org/x/ratelimit is close, it's just missing an Available() method we need to calculate saturation for metrics reporting. Opened https://github.com/golang/go/issues/18233

**Pull request merged**